src/utils/print_helpers.hpp

```diff
@@ -49,6 +50,23 @@ std::ostream &operator<<(std::ostream &in, const std::map<K, V> &map) {
   return in;
 }
 
+template <typename K, typename V>
+std::ostream &operator<<(std::ostream &in, const std::unordered_map<K, V> &map) {
+  in << "{";
+  bool first = true;
+  for (const auto &[a, b] : map) {
+    if (!first) {
+      in << ", ";
+    }
+    first = false;
+    in << a;
+    in << ": ";
+    in << b;
+  }
+  in << "}";
+  return in;
+}
+
```

Comment:
I think this is the same as the above one for `std::map`. The two can be done by the same function.

You can use template iterators to pass in a range, something like this: 
```
template <typename TIterator> void print(std::ostream &in, TIterator it, const TIterator end) {
in << "{";
  bool first = true;
  while (it != end) {
    const auto&[a,b] = *it;
    if (!first) {
      in << ", ";
    }
    first = false;
    in << a;
    in << ": ";
    in << b;
    ++it;
  }
  in << "}";
}
```

You can use concepts to restrict the types for iterators, but it is not necessary in my opinion.

Reply:
An important consideration of this is to work with the operator<< convention so that external libraries are able to also work with our types in a more graceful way. I don't mind the minor duplication here for common data structures since it's relatively well contained.